Common/util/bufferedstream.cpp

```diff
@@ -0,0 +1,125 @@
+//=============================================================================
+//
+// Adventure Game Studio (AGS)
+//
+// Copyright (C) 1999-2011 Chris Jones and 2011-20xx others
+// The full list of copyright holders can be found in the Copyright.txt
+// file, which is part of this source code distribution.
+//
+// The AGS source code is provided under the Artistic License 2.0.
+// A copy of this license can be found in the file License.txt and at
+// http://www.opensource.org/licenses/artistic-license-2.0.php
+//
+//=============================================================================
+
+#include "util/bufferedstream.h"
+
+#include <stdexcept>
+
+#include <cstring>
+#include <algorithm>
+
+#include "util/stdio_compat.h"
+#include "util/string.h"
+
+namespace AGS {
+namespace Common {
+
+BufferedStream::BufferedStream(const String &file_name, FileOpenMode open_mode, FileWorkMode work_mode, DataEndianess stream_endianess)
+        : FileStream(file_name, open_mode, work_mode, stream_endianess), buffer_(BufferStreamSize), bufferPosition_(0), position_(0)
+{
+    if (FileStream::Seek(0, kSeekEnd) == false) {
+        throw std::runtime_error("Error determining stream end.");
+    }
+
+    end_ = FileStream::GetPosition();
+    if (end_ == -1) {
+        throw std::runtime_error("Error determining stream end.");
+    }
+
+    if (FileStream::Seek(0, kSeekBegin) == false) {
+        throw std::runtime_error("Error determining stream end.");
+    }
+
+    buffer_.resize(0);
+
+}
+
+void BufferedStream::FillBufferFromPosition(soff_t position) {
+    FileStream::Seek(position, kSeekBegin);
+
+    buffer_.resize(BufferStreamSize);
+    auto sz = FileStream::Read(buffer_.data(), BufferStreamSize);
+    buffer_.resize(sz);
+
+	bufferPosition_ = position;
+}
+
+bool BufferedStream::EOS() const {
+    return position_ == end_;
+}
+
+soff_t BufferedStream::GetPosition() const {
+    return position_;
+}
+
+size_t BufferedStream::Read(void *toBuffer, size_t toSize) {
+    auto to = static_cast<char *>(toBuffer);
+
+    while(toSize > 0) {
+        if (position_ < bufferPosition_ || position_ >= bufferPosition_ + buffer_.size()) {
+            FillBufferFromPosition(position_);
+            if (buffer_.size() <= 0) { break; } // reached EOS
+        }
+
+        auto bufferOffset = position_ - bufferPosition_;
+        auto bytesLeft = buffer_.size() - bufferOffset;
+        auto chunkSize = std::min<size_t>(bytesLeft, toSize);
+
+        std::memcpy(to, buffer_.data()+bufferOffset, chunkSize);
+
+        to += chunkSize;
+        position_ += chunkSize;
+        toSize -= chunkSize;
+    }
+
+    return to - toBuffer;
+}
+
+int32_t BufferedStream::ReadByte() {
+    uint8_t ch;
+    auto bytesRead = Read(&ch, 1);
+    if (bytesRead != 1) { return EOF; }
+    return ch;
+}
+
+size_t BufferedStream::Write(const void *buffer, size_t size) { 
+    FileStream::Seek(position_, kSeekBegin);
+    auto sz = FileStream::Write(buffer, size);
+	position_ += sz;
```

Comment:
The code style here is contradicting (different identation for instance), and also not compying to our existing code style convention:
https://github.com/adventuregamestudio/ags/wiki/AGS-Coding-Conventions-(Cpp)

Are you planning on moving towards this codestyle from now on?

Reply:
I started doing this in places because I was getting style warnings. Identifiers that start with `_` are reserved in C++ in certain circumstances. 

But you're right, I'll review the code style again and try to stick with it. I'm not suggesting we run a style formatter tool over the whole project, but it would be useful for new code.

Reply:
Regarding underscore, we checked this back in 2012, and we found that having class member variables starting with underscore should be okay.

From https://stackoverflow.com/questions/228783/what-are-the-rules-about-using-an-underscore-in-a-c-identifier

> Reserved in any scope, including for use as implementation macros:
identifiers beginning with an underscore followed immediately by an uppercase letter
identifiers containing adjacent underscores (or "double underscore")
Reserved in the global namespace:
identifiers beginning with an underscore

Reply:
What style warnings were you getting? I've seen quite a few programs that use this naming style.

Reply:
If that would be a real issue we may change it of course, but may need to work out a convenient naming style for protected class fields.

Reply:
Hah I just read that same page. But just the fact that prepended underscores are reserved in some situations suggests we should avoid that for consistency.

The warnings were when using clang and turning all warnings on. It specifically complained about our `__` at the start of our header guards for example.

It's all good, I was trying a thing, I can revert in this pull request and future changes.

Reply:
> The warnings were when using clang and turning all warnings on. It specifically complained about our __ at the start of our header guards for example.

I did not realize that. We may change that removing leading underscores.